src/ImageManipulation.php

```diff
@@ -840,6 +841,18 @@ public function isHeight($height)
         return $this->getHeight() === $height;
     }
 
+    /**
+     * Wrapper for manipulate() that creates a variant file with a different extension than the original file.
+     *
+     * @return DBFile The manipulated file
+     */
+    public function manipulateExtension(string $newExtension, callable $callback)
```

Comment:
Note we can't add an explicit return typehint here because we're just using `return $this->manipulate(/**/);` which _should_ return `DBFile` but _could_ be overridden to return any arbitrary thing.

Reply:
Not sure that's a valid concern?  Who in the real world would subclass ImageManipulation and change manipulate()?  And even if they did they can just change the return type on this method?  We don't normally care about project override core classes. Am I missing something here?

Seems like return type should be `: ?DBFile` since manipulate() can return `null`

Reply:
It is a valid concern, because we have a whole public API document that assumes people will change anything that's public. The whole point is that we should assume people will change anything that's public, and that way we never run into people saying "You broke this public API that I was using in a bizarre but totally semver-friendly way!"

> And even if they did they can just change the return type on this method?

They can't just change the return type of this method if the return type they want isn't covariant (or contravariant, whichever the right one is here) with the return type we specify.

> We don't normally care about project override core classes. Am I missing something here?

We have a whole [definition of public API](https://docs.silverstripe.org/en/5/project_governance/public_api/) which assumes people will override core classes.

> Seems like return type should be : ?DBFile since manipulate() can return null

Updated the PHPDoc for each of the manipulate methods to reflect this.

Reply:
OK make sense now, it's purely because manipulate() doesn't a have a strong return type - which you'd already mentioned :-)